### PR TITLE
search: allow wildcards

### PR DESCRIPTION
**Pull request by vrothberg**
Allow trailing wildcards in the search term.  Note that not all
registries support wildcards and it may only work with v1 registries.
Hence, we just strip `*` before parsing the domain of the input.

Fixes: bugzilla.redhat.com/show_bug.cgi?id=1846629
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

@mtrmac @rhatdan @TomSweeneyRedHat  @QiWang19 PTAL

### CODE REVIEW

**Comment by vrothberg** [comment]
@baude @mheon this would be a nice to have for v2, assuming this PR is the right approach.

**Comment by TomSweeneyRedHat** [comment]
Some of the tests don't look happy.

**Comment by vrothberg** [comment]
> parse it manually, using whatever the semantics is 

That is the crux. I don't see a way to support arbitrary input. Especially with the search-registries. We need to extract a the domain/registry from the input, but if we support arbitrary input we just can't do that.

If we limit wildcard-support to trailing ones, we don't can keep the current parsing logic and support most of the use cases (at least all in the BZ).

**Comment by vrothberg** [comment]
Marked as WIP until we figured out if trailing-only would meet the customer demands.

**Comment by mtrmac** [comment]
> > parse it manually, using whatever the semantics is
> 
> That is the crux. I don't see a way to support arbitrary input. Especially with the search-registries. We need to extract a the domain/registry from the input, but if we support arbitrary input we just can't do that.

I didn’t mean “whatever the semantics of the remote registry wildcards is”, just “whatever the semantics of the argument to `podman search` is”.

Looking at https://github.com/containers/libpod/blob/88ee5d4f5f8ba6afabbb921db7b0ba7e92d59148/libpod/image/search.go#L71-L78 , we parse the input string into (`registry`, `term`), and only `term` is passed to the registry’s API (and thus has unknown semantics).

The semantics of the split into registry+term is entirely within Podman’s control. Isn’t what we want basically a “registry is the part before the first `/` if any”? I suppose we should have unit tests for all the affected cases, but I don’t think it’s much more than that.

**Comment by vrothberg** [comment]
Brilliant idea to consider a registry before `/`, @mtrmac!

**Comment by mtrmac** [comment]
Looking into this a bit further, the tests of `reference.Parse` do suggest that `foo/bar` is parsed as “domain == `foo`”.

OTOH there is a difference in error handling/format enforcement, for inputs like 'what#is/this\even'.

**Comment by vrothberg** [comment]
> Looking into this a bit further, the tests of `reference.Parse` do suggest that `foo/bar` is parsed as “domain == `foo`”.
> 
> OTOH there is a difference in error handling/format enforcement, for inputs like 'what#is/this\even'.

The podman-search behaviour is quite curious in some cases already:
```
~ $ podman search registry.redhat.io
INDEX       NAME                                               DESCRIPTION                                       STARS   OFFICIAL   AUTOMATED
docker.io   docker.io/knqyf263/registry.redhat.io-ubi7                                                           0                  
docker.io   docker.io/abdennour/rhel                           RHEL based on registry.redhat.io/ubi8/ubi-in...   0                  
docker.io   docker.io/chengkuan/amq-streams-kafka-connect-23   Kafka MongoDB Connect based on registry.redh...   0                  
```

It really is quite challenging for parsing. Docker has such a nice life as it only normalizes to docker.io.

File: libpod/image/search.go [hunk 1]
```diff
@@ -64,11 +64,11 @@ type SearchFilter struct {
 // SearchImages searches images based on term and the specified SearchOptions
 // in all registries.
 func SearchImages(term string, options SearchOptions) ([]SearchResult, error) {
-	// Check if search term has a registry in it
-	registry, err := sysreg.GetRegistry(term)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error getting registry from %q", term)
+	registry := ""
```
Comment (vrothberg) [inline libpod/image/search.go:67]: @mtrmac, does that reflect what you had in mind?

Comment (mtrmac) [reply libpod/image/search.go:67]: Yes, exactly (but I haven’t done the legwork to check that there are enough tests for the relevant input formats to know I’m not missing a corner case).

File: libpod/image/search.go [hunk 2]
```diff
@@ -64,11 +64,11 @@ type SearchFilter struct {
...
+	if spl := strings.SplitN(term, "/", 2); len(spl) > 1 {
+		registry = spl[0]
 	}
+
 	if registry != "" {
 		term = term[len(registry)+1:]
 	}
```
Comment (mtrmac) [inline libpod/image/search.go:74]: (Absolutely non-blocking: It feel be a tiny bit cleaner to me if this used `spl[1]` in the condition above instead of an independent parsing code — but AFAICS it works exactly the same.)

**Comment by vrothberg** [comment]
Rebased on master to unblock CI
